// advanced-cropper/node/service/helpers.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../utils/index.js');
var utils = require('./utils.js');
var sizeRestrictions = require('./sizeRestrictions.js');

function isInitializedState(state) {
    return Boolean(state && state.visibleArea && state.coordinates);
}
function getAreaSizeRestrictions(state, settings) {
    return sizeRestrictions.calculateAreaSizeRestrictions(state, settings);
}
function getAreaPositionRestrictions(state, settings) {
    return index.isFunction(settings.areaPositionRestrictions)
        ? settings.areaPositionRestrictions(state, settings)
        : settings.areaPositionRestrictions;
}
function getSizeRestrictions(state, settings) {
    return sizeRestrictions.calculateSizeRestrictions(state, settings);
}
function getPositionRestrictions(state, settings) {
    return index.isFunction(settings.positionRestrictions)
        ? settings.positionRestrictions(state, settings)
        : settings.positionRestrictions;
}
function getCoefficient(state) {
    return state.visibleArea ? state.visibleArea.width / state.boundary.width : 0;
}
function getStencilCoordinates(state) {
    if (isInitializedState(state)) {
        var _a = state.coordinates, width = _a.width, height = _a.height, left = _a.left, top_1 = _a.top;
        var coefficient = getCoefficient(state);
        return {
            width: width / coefficient,
            height: height / coefficient,
            left: (left - state.visibleArea.left) / coefficient,
            top: (top_1 - state.visibleArea.top) / coefficient,
        };
    }
    else {
        return index.emptyCoordinates();
    }
}
function getAspectRatio(state, settings) {
    return utils.createAspectRatio(index.isFunction(settings.aspectRatio) ? settings.aspectRatio(state, settings) : settings.aspectRatio);
}
function getDefaultCoordinates(state, settings) {
    return index.isFunction(settings.defaultCoordinates)
        ? settings.defaultCoordinates(state, settings)
        : settings.defaultCoordinates;
}
function getDefaultVisibleArea(state, settings) {
    return index.isFunction(settings.defaultVisibleArea)
        ? settings.defaultVisibleArea(state, settings)
        : settings.defaultVisibleArea;
}
function getTransformedImageSize(state) {
    if (state.imageSize && state.imageSize.width && state.imageSize.height) {
        return utils.rotateSize(state.imageSize, state.transforms.rotate);
    }
    else {
        return {
            width: 0,
            height: 0,
        };
    }
}
function getMinimumSize(state) {
    // The magic number is the approximation of the handler size
    // Temporary solution that should be improved in the future
    return state.coordinates
        ? Math.min(state.coordinates.width, state.coordinates.height, 20 * getCoefficient(state))
        : 1;
}
function getRoundedCoordinates(state, settings) {
    if (isInitializedState(state)) {
        var sizeRestrictions = getSizeRestrictions(state, settings);
        var positionRestrictions = getPositionRestrictions(state, settings);
        var roundCoordinates = {
            width: Math.round(state.coordinates.width),
            height: Math.round(state.coordinates.height),
            left: Math.round(state.coordinates.left),
            top: Math.round(state.coordinates.top),
        };
        if (roundCoordinates.width > sizeRestrictions.maxWidth) {
            roundCoordinates.width = Math.floor(state.coordinates.width);
        }
        else if (roundCoordinates.width < sizeRestrictions.minWidth) {
            roundCoordinates.width = Math.ceil(state.coordinates.width);
        }
        if (roundCoordinates.height > sizeRestrictions.maxHeight) {
            roundCoordinates.height = Math.floor(state.coordinates.height);
        }
        else if (roundCoordinates.height < sizeRestrictions.minHeight) {
            roundCoordinates.height = Math.ceil(state.coordinates.height);
        }
        return utils.moveToPositionRestrictions(roundCoordinates, positionRestrictions);
    }
    else {
        return null;
    }
}
function isConsistentState(state, settings) {
    if (isInitializedState(state)) {
        return (!utils.getBrokenRatio(utils.ratio(state.coordinates), getAspectRatio(state, settings)) &&
            utils.isConsistentSize(state.visibleArea, getAreaSizeRestrictions(state, settings)) &&
            utils.isConsistentSize(state.coordinates, getSizeRestrictions(state, settings)) &&
            utils.isConsistentPosition(state.visibleArea, getAreaPositionRestrictions(state, settings)) &&
            utils.isConsistentPosition(state.coordinates, getPositionRestrictions(state, settings)));
    }
    else {
        return true;
    }
}

exports.getAreaPositionRestrictions = getAreaPositionRestrictions;
exports.getAreaSizeRestrictions = getAreaSizeRestrictions;
exports.getAspectRatio = getAspectRatio;
exports.getCoefficient = getCoefficient;
exports.getDefaultCoordinates = getDefaultCoordinates;
exports.getDefaultVisibleArea = getDefaultVisibleArea;
exports.getMinimumSize = getMinimumSize;
exports.getPositionRestrictions = getPositionRestrictions;
exports.getRoundedCoordinates = getRoundedCoordinates;
exports.getSizeRestrictions = getSizeRestrictions;
exports.getStencilCoordinates = getStencilCoordinates;
exports.getTransformedImageSize = getTransformedImageSize;
exports.isConsistentState = isConsistentState;
exports.isInitializedState = isInitializedState;
