// @ggj/utils/utils/user.js
"use strict";
function developerId(oldDevUserId) {
    if (!oldDevUserId) {
        return 0;
    }
    // Exhibitor members- 出品者会員
    if (oldDevUserId < 10000) {
        return oldDevUserId + 110000;
    }
    // No mapping required for new members - 新規会員のためマッピング不要
    if (oldDevUserId >= 110000 && oldDevUserId < 200000) {
        return oldDevUserId;
    }
    // regular sellers members - 通常会員から出品者へ
    if (oldDevUserId >= 200000 && oldDevUserId < 300000) {
        return oldDevUserId - 200000;
    }
    // affiliates Seller - アフィリエイターから出品者
    if (oldDevUserId >= 400000 && oldDevUserId < 500000) {
        return oldDevUserId - 300000;
    }
    if (oldDevUserId > 600000) {
        return oldDevUserId;
    }
    return 0;
}
/**
 * Get user id in `asp._info_devuser`
 *
 * @param {number} userId
 * @returns {number}
 * @public
 */
function oldDeveloperId(userId) {
    const uid = parseInt(userId);
    if (uid > 140000) {
        return uid;
    }
    if (uid > 130000) {
        return 0;
    }
    if (uid > 120000) {
        return uid + 300000;
    }
    if (uid > 110000) {
        return uid - 110000;
    }
    return uid + 200000;
}
/**
 * Get user id from `asp._info_affuser`
 *
 * @param {number} oldPartnerId
 * @returns {number}
 * @public
 */
function partnerId(oldPartnerId) {
    const oldId = parseInt(oldPartnerId);
    if (oldId == 0) {
        return 0;
    }
    if (oldId > 600000) {
        return oldId;
    }
    if (oldId < 100000) {
        return 120000 + oldId;
    }
    if (oldId > 140000 && oldId < 200000) {
        return oldId;
    }
    if (oldId > 300000 && oldId < 400000) {
        return oldId - 300000;
    }
    if (oldId > 500000 && oldId < 600000) {
        return oldId - 400000;
    }
    return 0;
}
/**
 * Get user id in `asp._info_affuser`
 *
 * @param {number} userId
 * @returns {number}
 * @public
 */
function oldPartnerId(userId) {
    const uid = parseInt(userId);
    if (uid > 140000) {
        return uid;
    }
    if (uid > 130000) {
        return 0;
    }
    if (uid > 120000) {
        return uid - 120000;
    }
    if (uid > 110000) {
        return uid + 400000;
    }
    return uid + 300000;
}
/**
 * Get user id in `asp._ad_maruser`
 *
 * @param {number} userId
 * @returns {number}
 * @public
 */
function oldMarchantId(userId) {
    const uid = parseInt(userId);
    if (uid > 140000) {
        return uid;
    }
    if (uid > 130000) {
        return uid - 130000;
    }
    if (uid > 120000) {
        return 0;
    }
    if (uid > 110000) {
        return uid;
    }
    return 0;
}
/**
 * Get user id from `asp._ad_maruser`
 *
 * @param {Number} oldMarchantUserId
 * @returns {Number}
 * @public
 */
function marchantUserId(oldMarchantUserId) {
    const omuid = parseInt(oldMarchantUserId);
    if (omuid == 0) {
        return 0;
    }
    if (omuid < 100000) {
        return 130000 + omuid;
    }
    if (omuid > 140000) {
        return omuid;
    }
    return 0;
}
/**
 * Convert new user id to old user id
 */
function getOldUserId(userId) {
    if (userId >= 140000 || userId < 110000) {
        return userId;
    }
    if (userId >= 130000) {
        return userId - 130000;
    }
    if (userId >= 120000) {
        return userId - 120000;
    }
    if (userId >= 110000) {
        return userId - 110000;
    }
}
module.exports = {
    developerId,
    oldDeveloperId,
    partnerId,
    oldPartnerId,
    marchantUserId,
    oldMarchantId,
    getOldUserId,
};
