// advanced-cropper/node/state/createState.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('tslib');
var helpers = require('../service/helpers.js');
var index = require('../types/index.js');
var setCoordinates = require('./setCoordinates.js');
var setVisibleArea = require('./setVisibleArea.js');

function createState(options, settings) {
    var _a, _b;
    var boundary = options.boundary, imageSize = options.imageSize, transforms = options.transforms, priority = options.priority;
    var state = {
        boundary: {
            width: boundary.width,
            height: boundary.height,
        },
        imageSize: {
            width: imageSize.width,
            height: imageSize.height,
        },
        transforms: {
            rotate: (transforms === null || transforms === void 0 ? void 0 : transforms.rotate) || 0,
            flip: {
                horizontal: ((_a = transforms === null || transforms === void 0 ? void 0 : transforms.flip) === null || _a === void 0 ? void 0 : _a.horizontal) || false,
                vertical: ((_b = transforms === null || transforms === void 0 ? void 0 : transforms.flip) === null || _b === void 0 ? void 0 : _b.vertical) || false,
            },
        },
        visibleArea: null,
        coordinates: null,
    };
    if (priority === index.Priority.visibleArea) {
        state = setVisibleArea.setVisibleArea(state, settings, helpers.getDefaultVisibleArea(state, settings), false);
        state = setCoordinates.setCoordinates(state, settings, helpers.getDefaultCoordinates(state, settings), setCoordinates.SetCoordinatesMode.limit);
    }
    else {
        state = setCoordinates.setCoordinates(state, settings, helpers.getDefaultCoordinates(state, settings), setCoordinates.SetCoordinatesMode.unsafe);
        state = setVisibleArea.setVisibleArea(state, settings, helpers.getDefaultVisibleArea(state, settings), true);
    }
    return state;
}

exports.createState = createState;
