// advanced-cropper/node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index$6 = require('./utils/index.js');
var utils = require('./service/utils.js');
var sizeRestrictions = require('./service/sizeRestrictions.js');
var helpers = require('./service/helpers.js');
var copyState = require('./state/copyState.js');
var index$5 = require('./types/index.js');
var setCoordinates = require('./state/setCoordinates.js');
var setVisibleArea = require('./state/setVisibleArea.js');
var createState = require('./state/createState.js');
var moveCoordinates = require('./state/moveCoordinates.js');
var resizeCoordinates = require('./state/resizeCoordinates.js');
var setBoundary = require('./state/setBoundary.js');
var transformImage = require('./state/transformImage.js');
var reconcileState = require('./state/reconcileState.js');
var approximateSize = require('./service/approximateSize.js');
var fitCoordinates = require('./service/fitCoordinates.js');
var fitVisibleArea = require('./service/fitVisibleArea.js');
var normalize = require('./service/normalize.js');
var imageTransforms = require('./service/imageTransforms.js');
var isEqualState = require('./service/isEqualState.js');
var moveCoordinatesAlgorithm = require('./algorithms/moveCoordinatesAlgorithm.js');
var resizeCoordinatesAlgorithm = require('./algorithms/resizeCoordinatesAlgorithm.js');
var rotateImageAlgorithm = require('./algorithms/rotateImageAlgorithm.js');
var flipImageAlgorithm = require('./algorithms/flipImageAlgorithm.js');
var transformImageAlgorithm = require('./algorithms/transformImageAlgorithm.js');
var defaultPositionRestrictions = require('./defaults/defaultPositionRestrictions.js');
var defaultVisibleArea = require('./defaults/defaultVisibleArea.js');
var defaultSize = require('./defaults/defaultSize.js');
var defaultSizeRestrictions = require('./defaults/defaultSizeRestrictions.js');
var defaultPosition = require('./defaults/defaultPosition.js');
var defaultAreaPositionRestrictions = require('./defaults/defaultAreaPositionRestrictions.js');
var defaultAreaSizeRestrictions = require('./defaults/defaultAreaSizeRestrictions.js');
var defaultBoundary = require('./defaults/defaultBoundary.js');
var defaultStencilConstraints = require('./defaults/defaultStencilConstraints.js');
var index = require('./defaults/index.js');
var AbstractCropper = require('./instance/AbstractCropper.js');
var CropperInstance = require('./instance/CropperInstance.js');
var index$2 = require('./animation/index.js');
var index$3 = require('./canvas/index.js');
var index$1 = require('./boundary/index.js');
var index$4 = require('./image/index.js');



exports.debounce = index$6.debounce;
exports.deepClone = index$6.deepClone;
exports.deepCompare = index$6.deepCompare;
exports.distance = index$6.distance;
exports.emptyCoordinates = index$6.emptyCoordinates;
exports.getCloserAngle = index$6.getCloserAngle;
exports.getDirectionNames = index$6.getDirectionNames;
exports.getOptions = index$6.getOptions;
exports.isArray = index$6.isArray;
exports.isArrayBufferLike = index$6.isArrayBufferLike;
exports.isBlob = index$6.isBlob;
exports.isCardinalDirection = index$6.isCardinalDirection;
exports.isCrossOriginURL = index$6.isCrossOriginURL;
exports.isDataUrl = index$6.isDataUrl;
exports.isFunction = index$6.isFunction;
exports.isGreater = index$6.isGreater;
exports.isLocal = index$6.isLocal;
exports.isLower = index$6.isLower;
exports.isMouseEvent = index$6.isMouseEvent;
exports.isNaN = index$6.isNaN;
exports.isNumber = index$6.isNumber;
exports.isNumeric = index$6.isNumeric;
exports.isObject = index$6.isObject;
exports.isOrdinalDirection = index$6.isOrdinalDirection;
exports.isRoughlyEqual = index$6.isRoughlyEqual;
exports.isString = index$6.isString;
exports.isTouchEvent = index$6.isTouchEvent;
exports.isUndefined = index$6.isUndefined;
exports.isWheelEvent = index$6.isWheelEvent;
exports.parseNumber = index$6.parseNumber;
exports.promiseTimeout = index$6.promiseTimeout;
exports.sign = index$6.sign;
exports.applyDirections = utils.applyDirections;
exports.applyMove = utils.applyMove;
exports.applyScale = utils.applyScale;
exports.aspectRatioIntersection = utils.aspectRatioIntersection;
exports.coordinatesToPositionRestrictions = utils.coordinatesToPositionRestrictions;
exports.createAspectRatio = utils.createAspectRatio;
exports.diff = utils.diff;
exports.fitToPositionRestrictions = utils.fitToPositionRestrictions;
exports.fitToSizeRestrictions = utils.fitToSizeRestrictions;
exports.getBrokenRatio = utils.getBrokenRatio;
exports.getCenter = utils.getCenter;
exports.getCloserSize = utils.getCloserSize;
exports.getIntersections = utils.getIntersections;
exports.getTransitionStyle = utils.getTransitionStyle;
exports.inverseMove = utils.inverseMove;
exports.isConsistentPosition = utils.isConsistentPosition;
exports.isConsistentSize = utils.isConsistentSize;
exports.maxScale = utils.maxScale;
exports.mergePositionRestrictions = utils.mergePositionRestrictions;
exports.minScale = utils.minScale;
exports.moveToPositionRestrictions = utils.moveToPositionRestrictions;
exports.positionToSizeRestrictions = utils.positionToSizeRestrictions;
exports.ratio = utils.ratio;
exports.resizeToSizeRestrictions = utils.resizeToSizeRestrictions;
exports.rotatePoint = utils.rotatePoint;
exports.rotateSize = utils.rotateSize;
exports.sizeDistance = utils.sizeDistance;
exports.calculateAreaSizeRestrictions = sizeRestrictions.calculateAreaSizeRestrictions;
exports.calculateSizeRestrictions = sizeRestrictions.calculateSizeRestrictions;
exports.mergeSizeRestrictions = sizeRestrictions.mergeSizeRestrictions;
exports.reconcileSizeRestrictions = sizeRestrictions.reconcileSizeRestrictions;
exports.getAreaPositionRestrictions = helpers.getAreaPositionRestrictions;
exports.getAreaSizeRestrictions = helpers.getAreaSizeRestrictions;
exports.getAspectRatio = helpers.getAspectRatio;
exports.getCoefficient = helpers.getCoefficient;
exports.getDefaultCoordinates = helpers.getDefaultCoordinates;
exports.getDefaultVisibleArea = helpers.getDefaultVisibleArea;
exports.getMinimumSize = helpers.getMinimumSize;
exports.getPositionRestrictions = helpers.getPositionRestrictions;
exports.getRoundedCoordinates = helpers.getRoundedCoordinates;
exports.getSizeRestrictions = helpers.getSizeRestrictions;
exports.getStencilCoordinates = helpers.getStencilCoordinates;
exports.getTransformedImageSize = helpers.getTransformedImageSize;
exports.isConsistentState = helpers.isConsistentState;
exports.isInitializedState = helpers.isInitializedState;
exports.copyState = copyState.copyState;
Object.defineProperty(exports, 'ImageRestriction', {
	enumerable: true,
	get: function () {
		return index$5.ImageRestriction;
	}
});
Object.defineProperty(exports, 'Priority', {
	enumerable: true,
	get: function () {
		return index$5.Priority;
	}
});
Object.defineProperty(exports, 'SetCoordinatesMode', {
	enumerable: true,
	get: function () {
		return setCoordinates.SetCoordinatesMode;
	}
});
exports.setCoordinates = setCoordinates.setCoordinates;
exports.setVisibleArea = setVisibleArea.setVisibleArea;
exports.createState = createState.createState;
exports.moveCoordinates = moveCoordinates.moveCoordinates;
exports.resizeCoordinates = resizeCoordinates.resizeCoordinates;
exports.setBoundary = setBoundary.setBoundary;
exports.transformImage = transformImage.transformImage;
exports.reconcileState = reconcileState.reconcileState;
exports.approximateSize = approximateSize.approximateSize;
exports.fitCoordinates = fitCoordinates.fitCoordinates;
exports.fitVisibleArea = fitVisibleArea.fitVisibleArea;
exports.fillMoveDirections = normalize.fillMoveDirections;
exports.fillResizeDirections = normalize.fillResizeDirections;
exports.normalizeCenter = normalize.normalizeCenter;
exports.normalizeFlip = normalize.normalizeFlip;
exports.normalizeImageTransform = normalize.normalizeImageTransform;
exports.normalizeMoveDirections = normalize.normalizeMoveDirections;
exports.normalizeResizeDirections = normalize.normalizeResizeDirections;
exports.touchesToImageTransform = imageTransforms.touchesToImageTransform;
exports.wheelEventToImageTransform = imageTransforms.wheelEventToImageTransform;
exports.isEqualState = isEqualState.isEqualState;
exports.moveCoordinatesAlgorithm = moveCoordinatesAlgorithm.moveCoordinatesAlgorithm;
exports.fitDirections = resizeCoordinatesAlgorithm.fitDirections;
exports.resizeCoordinatesAlgorithm = resizeCoordinatesAlgorithm.resizeCoordinatesAlgorithm;
exports.rotateImageAlgorithm = rotateImageAlgorithm.rotateImageAlgorithm;
exports.flipImageAlgorithm = flipImageAlgorithm.flipImageAlgorithm;
exports.transformImageAlgorithm = transformImageAlgorithm.transformImageAlgorithm;
exports.defaultPositionRestrictions = defaultPositionRestrictions.defaultPositionRestrictions;
exports.defaultVisibleArea = defaultVisibleArea.defaultVisibleArea;
exports.defaultSize = defaultSize.defaultSize;
exports.pixelsRestrictions = defaultSizeRestrictions.pixelsRestrictions;
exports.retrieveSizeRestrictions = defaultSizeRestrictions.retrieveSizeRestrictions;
exports.defaultPosition = defaultPosition.defaultPosition;
exports.defaultAreaPositionRestrictions = defaultAreaPositionRestrictions.defaultAreaPositionRestrictions;
exports.defaultAreaSizeRestrictions = defaultAreaSizeRestrictions.defaultAreaSizeRestrictions;
exports.fillBoundary = defaultBoundary.fillBoundary;
exports.fitBoundary = defaultBoundary.fitBoundary;
exports.defaultStencilConstraints = defaultStencilConstraints.defaultStencilConstraints;
exports.createDefaultSettings = index.createDefaultSettings;
exports.withDefaultAreaPositionRestrictions = index.withDefaultAreaPositionRestrictions;
exports.withDefaultAreaSizeRestrictions = index.withDefaultAreaSizeRestrictions;
exports.withDefaultPositionRestrictions = index.withDefaultPositionRestrictions;
exports.withDefaultSizeRestrictions = index.withDefaultSizeRestrictions;
exports.AbstractCropper = AbstractCropper.AbstractCropper;
exports.CropperInstance = CropperInstance.CropperInstance;
exports.Animation = index$2.Animation;
exports.drawCroppedArea = index$3.drawCroppedArea;
exports.prepareSource = index$3.prepareSource;
exports.updateCanvas = index$3.updateCanvas;
exports.stretchCropperBoundary = index$1.stretchCropperBoundary;
exports.stretchPreviewBoundary = index$1.stretchPreviewBoundary;
exports.createImage = index$4.createImage;
exports.getBackgroundStyle = index$4.getBackgroundStyle;
exports.getImageStyle = index$4.getImageStyle;
exports.getPreviewStyle = index$4.getPreviewStyle;
exports.getStyleTransforms = index$4.getStyleTransforms;
exports.loadImage = index$4.loadImage;
